### PR TITLE
Add EntityResolver and DocType event

### PR DESCRIPTION
**Pull request by pigeonhands**
fixes https://github.com/tafia/quick-xml/issues/581

+ Added `PayloadEvent::DocType`
+ Added an `EntityResolver`  so that unknown entities can be resolved.
+ Changed `unsescape_` functions to use `FnMut` instead of `Fn` for `resolve_entity` parameters.

### CODE REVIEW

**Comment by dralley** [comment]
Please add an entry to the changelog

**Comment by pigeonhands** [comment]
@dralley Done

File: src/de/mod.rs [hunk 1]
```diff
@@ -2167,12 +2184,12 @@ where
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
 /// A structure that deserializes XML into Rust values.
-pub struct Deserializer<'de, R>
+pub struct Deserializer<'de, R, S: EntityResolver = DefaultEntityResolver>
```
Comment (Mingun) [inline src/de/mod.rs:2187]: Let's use consistent naming
```suggestion
pub struct Deserializer<'de, R, E: EntityResolver = DefaultEntityResolver>
```

File: src/escapei.rs [hunk 2]
```diff
@@ -159,11 +159,11 @@ pub fn unescape(raw: &str) -> Result<Cow<str>, EscapeError> {
 /// [HTML5 escapes]: https://dev.w3.org/html5/html-author/charref
 pub fn unescape_with<'input, 'entity, F>(
     raw: &'input str,
-    resolve_entity: F,
+    mut resolve_entity: F,
 ) -> Result<Cow<'input, str>, EscapeError>
 where
     // the lifetime of the output comes from a capture or is `'static`
-    F: Fn(&str) -> Option<&'entity str>,
+    F: FnMut(&str) -> Option<&'entity str>,
```
Comment (Mingun) [inline src/escapei.rs:166]: After consulting with the [standard library](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map) I think, it's a good idea, but that change should be done in it's own commit and have it's own changelog entry.

File: src/resolver.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+/// An EntityResolver that always returns None.
+#[derive(Default, Copy, Clone)]
+pub struct DefaultEntityResolver;
+
+impl<'entity_out> EntityResolver for DefaultEntityResolver{
+    fn resolve<'entity>(&'entity self, _: &str) -> Option<&'entity str> {
+        None
+    }
+
+    fn capture(&mut self, _: BytesText){
```
Comment (Mingun) [inline src/resolver.rs:27]: Let's be consistent and put this method before the `resolve` method

File: src/resolver.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    /// `None` is returned if a suitable value can not be found.
+    /// In that case an [`Error::UnrecognizedSymbol`] will be returned.
+    fn resolve<'entity>(&'entity self, entity: &str) -> Option<&'entity str>;
+}
+
+/// An EntityResolver that always returns None.
+#[derive(Default, Copy, Clone)]
+pub struct DefaultEntityResolver;
+
+impl<'entity_out> EntityResolver for DefaultEntityResolver{
```
Comment (Mingun) [inline src/resolver.rs:22]:
```suggestion
impl EntityResolver for DefaultEntityResolver {
```

File: src/resolver.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    /// Called when an entity needs to be resolved.
+    ///
+    /// `None` is returned if a suitable value can not be found.
+    /// In that case an [`Error::UnrecognizedSymbol`] will be returned.
+    fn resolve<'entity>(&'entity self, entity: &str) -> Option<&'entity str>;
+}
+
+/// An EntityResolver that always returns None.
+#[derive(Default, Copy, Clone)]
+pub struct DefaultEntityResolver;
```
Comment (Mingun) [inline src/resolver.rs:20]:
```suggestion
pub struct NoEntityResolver;
```

File: src/resolver.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
+//! Entity resolver module
+
+use crate::events::BytesText;
+
+/// Used to resolve unknown entities while parsing
+pub trait EntityResolver {
```
Comment (Mingun) [inline src/resolver.rs:6]: Please add an example here as suggested in [my comment](https://github.com/tafia/quick-xml/issues/581#issuecomment-1488823175)

File: src/de/mod.rs [hunk 7]
```diff
@@ -1936,6 +1937,9 @@ pub enum PayloadEvent<'a> {
     /// Unescaped character data between `Start` and `End` element,
     /// stored in `<![CDATA[...]]>`.
     CData(BytesCData<'a>),
+    /// Unescaped character data between `Start` and `End` element,
+    /// stored in `<!DOCTYPE ... [...]]>`.
```
Comment (Mingun) [inline src/de/mod.rs:1941]:
```suggestion
    /// Document type definition data (DTD) stored in `<!DOCTYPE ...]>`.
```
Actually, I would like if you would synchronize documentation between `Event` and `PayloadEvent` (in it's own commit, do that first)

File: src/de/mod.rs [hunk 8]
```diff
@@ -1949,31 +1953,40 @@ impl<'a> PayloadEvent<'a> {
...
+struct XmlReader<'i, R: XmlRead<'i>, E: EntityResolver = DefaultEntityResolver> {
     /// A source of low-level XML events
     reader: R,
     /// Intermediate event, that could be returned by the next call to `next()`.
     /// If that is the `Text` event then leading spaces already trimmed, but
     /// trailing spaces is not. Before the event will be returned, trimming of
     /// the spaces could be necessary
     lookahead: Result<PayloadEvent<'i>, DeError>,
+
+    entity_resolver: E
```
Comment (Mingun) [inline src/de/mod.rs:1975]: Please add a doc comment explaining for what this field is used

**Review by Mingun — CHANGES_REQUESTED** [review]
Thanks for your work! I think PR will benefit if you make such changes: split it into three commits (rebase your branch and force push it):
- the first one that synchronizes documentation between variants of `Event` and `PayloadEvent`
- the second one which relaxes requirements to `resolve_entity` parameters from `Fn` to `FnMut`. That commit should relax requirements in all places and note this in the changelog
- the third one which adds an `EntityResolver` trait etc. I was too quick in my comment regarding the `resolve` method. Because `resolve_entity` are now would be `FnMut`, the trait could be defined as
  ```rust
  pub trait EntityResolver {
      fn capture(&mut self, doctype: BytesText);
      fn resolve(&mut self, entity: &str) -> Option<&str>;
  }
  ```
  Add tests for the new feature in this commit

File: src/escapei.rs (hunk 2 above)
Comment (dralley) [reply src/escapei.rs:166]: It's not immediately clear why it's necessary, as the docs [say](https://doc.rust-lang.org/std/ops/trait.FnMut.html):

> Use FnMut as a bound when you want to accept a parameter of function-like type and need to call it repeatedly, while allowing it to mutate state. If you don’t want the parameter to mutate state, use [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html) as a bound; if you don’t need to call it repeatedly, use [FnOnce](https://doc.rust-lang.org/std/ops/trait.FnOnce.html).

What state do you expect the entity resolution process to mutate?  Isn't it a read-only process?

Comment (Mingun) [reply src/escapei.rs:166]: Yes, it is just to provide flexibility and be consistent with similar std API, such as `Iterator::map`.

**Comment by pigeonhands** [comment]
> This is almost good, just need to polish some things.
> 
> Actually, the first commit did not what I want. I meant that you to change the doc strings of all matching `Event` and `PayloadEvent` variants to be the same by keeping the best description. And then in the third commit add a new variant to `PayloadEvent` (so it would be just a copy of the corresponding variant from `Event`)

I'm not sure what you mean by this. Should all the enum variants be the same between `PayloadEvent` and `Event` or just the docs?

**Comment by Mingun** [comment]
Just the docs, because `PayloadEvent` is a subset of possible `Event`s, and docs should be the same on same variants.

**Review by Mingun — APPROVED** [review]
I've updated the PR, with the following changes:
- Allow `EntityResolver::capture` method to fail
- Change the example on `EntityResolver` more close to possible real world implementation
- Make `resolver` module private and instead re-export trait and type in `de` module

Unfortunately, making `EntityResolver::resolve` taking `&mut self` leads to strange error, that the returned reference could not outlive the captured object in the closure. This is strange, because returning the same reference while resolver is read-only is fine. So if you wish to make lazy parsing in `resolve`, you should use `Cell` or similar.

@dralley, could you look at this once again? It seems good for me now and I plan to make another patch release after merging that.

File: tests/serde-de.rs [hunk 9]
```diff
@@ -6427,3 +6427,71 @@ mod borrow {
...
+
+    impl EntityResolver for TestEntityResolver {
+        fn capture(&mut self, doctype: quick_xml::events::BytesText) {
+            self.capture_called = true;
+            let as_cow = doctype.into_inner();
+            let str_doc = String::from_utf8_lossy(as_cow.as_ref());
+            assert_eq!(str_doc.as_ref(), r#"dict[ <!ENTITY unc "unclassified"> ]"#);
+        }
+
+        fn resolve<'entity>(&'entity self, entity: &str) -> Option<&'entity str> {
```
[the remark this answers is not in the document]
Comment (Mingun) [reply tests/serde-de.rs]: Change to `&mut` is not liked by rustc, so keep it as is

**Review by dralley — APPROVED** [review]
